[axonius][alert_and_incident] Add Axonius Alert and Incident datastream

muskan-agarwal26 opened the pull request:
## Proposed commit message

This new data stream collects alert findings and incidents data from the Axonius platform via the 
/api/v2/assets/ endpoint. Axonius fields are mapped to their 
corresponding ECS fields where possible.

Test samples were derived from live data samples, which were subsequently
sanitized.

API documentation: https://docs.axonius.com/docs/axonius-rest-api

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

## How to test this PR locally

To test the axonius package:

- Clone integrations repo.
- Install elastic package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/axonius directory.
- Run the following command to run tests.

> elastic-package test
```
2025/12/17 18:21:56  INFO New version is available - v0.117.1. Download from: https://github.com/elastic/elastic-package/releases/tag/v0.117.1
Run asset tests for the package
2025/12/17 18:21:57  INFO License text found in "/root/GITHUB/integrations/LICENSE.txt" will be included in package
--- Test results for package: axonius - START ---
╭─────────┬────────────────────┬───────────┬──────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM        │ TEST TYPE │ TEST NAME                                                        │ RESULT │ TIME ELAPSED │
├─────────┼────────────────────┼───────────┼──────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ axonius │                    │ asset     │ dashboard axonius-50841034-1314-4e26-8be3-47d38241c359 is loaded │ PASS   │      2.703µs │
│ axonius │                    │ asset     │ search axonius-852dd7b3-5b99-456b-8f00-9db6b8d6f4bd is loaded    │ PASS   │        481ns │
│ axonius │                    │ asset     │ search axonius-d3b2048c-016a-4ebb-9406-89687aa90aee is loaded    │ PASS   │        546ns │
│ axonius │ alert_and_incident │ asset     │ index_template logs-axonius.alert_and_incident is loaded         │ PASS   │        696ns │
│ axonius │ alert_and_incident │ asset     │ ingest_pipeline logs-axonius.alert_and_incident-0.1.0 is loaded  │ PASS   │        345ns │
╰─────────┴────────────────────┴───────────┴──────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run pipeline tests for the package
--- Test results for package: axonius - START ---
╭─────────┬────────────────────┬───────────┬───────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM        │ TEST TYPE │ TEST NAME                                         │ RESULT │ TIME ELAPSED │
├─────────┼────────────────────┼───────────┼───────────────────────────────────────────────────┼────────┼──────────────┤
│ axonius │ alert_and_incident │ pipeline  │ (ingest pipeline warnings test-alert-finding.log) │ PASS   │ 726.121672ms │
│ axonius │ alert_and_incident │ pipeline  │ (ingest pipeline warnings test-incident.log)      │ PASS   │ 712.669141ms │
│ axonius │ alert_and_incident │ pipeline  │ test-alert-finding.log                            │ PASS   │ 194.767219ms │
│ axonius │ alert_and_incident │ pipeline  │ test-incident.log                                 │ PASS   │ 209.348823ms │
╰─────────┴────────────────────┴───────────┴───────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run policy tests for the package
--- Test results for package: axonius - START ---
No test results
--- Test results for package: axonius - END   ---
Done
Run static tests for the package
--- Test results for package: axonius - START ---
╭─────────┬────────────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM        │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├─────────┼────────────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ axonius │ alert_and_incident │ static    │ Verify sample_event.json │ PASS   │ 220.387313ms │
╰─────────┴────────────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run system tests for the package
2025/12/17 18:22:05  INFO Installing package...
2025/12/17 18:22:05  INFO License text found in "/root/GITHUB/integrations/LICENSE.txt" will be included in package
2025/12/17 18:22:16  INFO Running test for data_stream "alert_and_incident" with configuration 'default'
2025/12/17 18:22:25  INFO Setting up independent Elastic Agent...
2025/12/17 18:22:39  INFO Setting up service...
2025/12/17 18:23:03  INFO Validating test case...
2025/12/17 18:23:04  INFO Tearing down service...
2025/12/17 18:23:05  INFO Write container logs to file: /root/GITHUB/integrations/build/container-logs/axonius-1765975985552405152.log
2025/12/17 18:23:08  INFO Tearing down agent...
2025/12/17 18:23:08  INFO Write container logs to file: /root/GITHUB/integrations/build/container-logs/elastic-agent-1765975988989458774.log
2025/12/17 18:23:36  INFO Uninstalling package...
--- Test results for package: axonius - START ---
╭─────────┬────────────────────┬───────────┬───────────┬────────┬───────────────╮
│ PACKAGE │ DATA STREAM        │ TEST TYPE │ TEST NAME │ RESULT │  TIME ELAPSED │
├─────────┼────────────────────┼───────────┼───────────┼────────┼───────────────┤
│ axonius │ alert_and_incident │ system    │ default   │ PASS   │ 47.688153488s │
╰─────────┴────────────────────┴───────────┴───────────┴────────┴───────────────╯
--- Test results for package: axonius - END   ---
Done
``` 

## Related issues

- Relates https://github.com/elastic/integrations/issues/15034
- Relates https://github.com/elastic/elasticsearch/pull/139512

## Screenshots
<img width="1919" height="913" alt="image (5)" src="https://github.com/user-attachments/assets/c254d523-7517-4137-a726-84e7c77f34fa" />
<img width="1919" height="916" alt="image (6)" src="https://github.com/user-attachments/assets/9abed0f7-58f5-4b4e-aca9-474d6de7b50f" />

muskan-crest committed:
Added alert and incident datastream in axonius

muskan-crest committed:
Added transform pipeline.

muskan-crest committed:
Made changes in cel code as suggested in exposure datastream

muskan-crest committed:
Updtaed test logs

muskan-crest committed:
Updated readme

chrisberkhout commented on packages/axonius/elasticsearch/transform/latest_alert_and_incident/transform.yml:
```diff
@@ -0,0 +1,37 @@
+# Use of "*" to use all namespaces defined.
```
It would be good to explain the transform in the README. I see it's mentioned that transforms are installed.

As a user I'd like to know that all data is periodically refetched and that's why there are latest transforms, and also the names of the aliases to access the latest data rather than the potentially duplicated data in the source data streams.

muskan-agarwal26 replied:
@chrisberkhout 
So far, we have been following the current approach across all integrations, where the README template is generated automatically. Is there any reference or documentation for the approach you’re suggesting so that I can implement the changes accordingly?

muskan-agarwal26 replied:
Still didn't find any idea about it.
In `ti_abusech`, it is mentioned for the ioc indicators.

chrisberkhout reviewed (commented):
Had a high-level look and I have a few questions about the approach.

I see that this is a PR to a feature branch rather than to main. That's fine, but it would have been helpful to mention that in the PR description and maybe one sentence about the scope of this PR and where it fits in the project.

muskan-agarwal26 commented on packages/axonius/img/axonius-alert-and-incident-dashboard.png:
Recently, we had got a comment on PR to apply filters on dashboard level, hence followed same here.

muskan-agarwal26 commented on packages/axonius/data_stream/alert_and_incident/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,180 @@
+      name: '{{ IngestPipeline "pipeline-incident" }}'
+      tag: pipeline-incident
+      if: >-
+        ctx.axonius?.alert_and_incident?.asset_type.contains('incidents')
```
Hi @chrisberkhout, there is one key difference across the data streams: the asset types passed within the endpoint vary for each stream. We can’t send all asset types in a single request because there are around 60 of them, each carrying a large volume of data, which could negatively impact API performance and cause latency.

One possible solution is to provide toggles that allow users to select which asset types they want to fetch. For example, the Exposure data stream may include 3 asset types, while the Compute data stream may include 10. If a user enables the Exposure toggle, only those 3 asset types would be fetched, and similarly for the other data streams.

muskan-agarwal26 commented on packages/axonius/data_stream/alert_and_incident/elasticsearch/ingest_pipeline/pipeline-alert-finding.yml:
```diff
@@ -0,0 +1,96 @@
+      field: json.event.data.alert_config_id
+      tag: rename_event_data_alert_config_id
+      target_field: axonius.alert_and_incident.event.data.alert_config_id
+      ignore_missing: true
```
`Also, I think the .event and .data prefixes are redundant. For this field, for example, axonius.alert_and_incident.alert_config_id is better (or axonius.alert.alert_config_id if we have an alert-specific data stream as discussed elsewhere).`

@chrisberkhout 
We follow this same mapping approach across the other data streams as well. Making a change here would require updating the others too—one is already merged and seven more are in review state. This would result in significant rework, so I suggest we skip this change for now. What do you think?

muskan-crest committed:
Resolved comments provided by @chrisberkhout.
1. Split alert_and_incident into alert_finding and incident.

chrisberkhout reviewed (commented):
Thanks for splitting the `alert_and_incident` data stream into `alert_finding` and `incident` data streams!

Regarding the other points from the last review, I checked each and made some comments. I think most of those still need some attention.

muskan-crest committed:
Resolved comment provided by @chrisberkhout.
1. Added transform info in readme.
2. Added filters on visual level, removed from dashboard level.
3. Removed event.data.* naming from custom mappings.
4. Added asset_type_list config param.

chrisberkhout reviewed (changes requested):
Please review the unresolved items from earlier reviews.

muskan-crest committed:
Resolved comment given by - @chrisberkhout.
1. Shifted transform filter to visual level.
2. Suggestions in readme and changelog.
